### DISCUSSION TITLE
A local-var can't be stored in a type

### DISCUSSION THREAD

**Discussion by anfelor** [koka-lang/koka#116] [General]
For example, this definition of a graph:
```
private struct graph<h> {
    nodes_ : local-var<h, vector<node>>
    num_edges_ : int
}
```
yields
```
graph.kk(49, 5): error: types do not match
  context      :     nodes_
  term         :     nodes_
  inferred type: vector<node>
  expected type: local-var<$h,vector<node>>
  because      : function returns values of different types
```
Is this expected behavior?

EDIT: One can use `ref` instead of `local-var` which makes updates of the vector a bit more cumbersome but works as expected.

**Comment by daanx** [comment]
Hi, good question/issue. I will try to document this more -- the `local-var` is an internal type used by `var` declarations; these are limited as their life-time is always limited to their lexical scope (checked through the `h` parameter). However, by limiting them we can give them the exact semantics of a `state` effect and create (in the future) more efficient implementations.

The `ref<h,a>` is the general type of references that are first-class and can have life times that exceed the lexical scope. An example of references is in the `samples/basic/garsia-wachs` example.

Hope this helps

**Reply by anfelor** [reply]
Thank you, yes, `ref` seems to be the better option. I originally used `local-var` because `([])` for vectors uses it.. maybe the function should also be provided for `ref`?

**Reply by daanx** [reply]
Ah, the vector assignment right; Yes, I think we need this for `ref` as well -- it needs some re-thinking now that Perceus is implemented; when vectors are unique (-ly owned) they can be updated in-place. That makes the need for ref's to vectors less needed perhaps? In particular updating an element in the vector that the ref points too. Not sure :-)
